### PR TITLE
Don't run optJumpThreading when fgCurBBEpochSize != (fgBBNumMax + 1)

### LINKED ISSUE #71599: Assertion failed 'i < BitSetTraits::GetSize(env)' in during 'Redundant branch opts'
```csharp
// Generated by Fuzzlyn v1.5 on 2022-07-03 16:41:26
// Run on Arm64 MacOS
// Seed: 15254214336696406464
// Reduced from 47.2 KiB to 1.6 KiB in 00:00:23
// Hits JIT assert in Release:
// Assertion failed 'i < BitSetTraits::GetSize(env)' in 'Program:Main(Fuzzlyn.ExecutionServer.IRuntime)' during 'Redundant branch opts' (IL size 260; hash 0xade6b36b; FullOpts)
// 
//     File: /Users/runner/work/1/s/src/coreclr/jit/bitsetasshortlong.h Line: 261
// 
public class C0
{
    public ushort F0;
    public ulong F1;
}

public class C1
{
    public C0 F0;
    public C0 F3;
    public C0 F6;
}

public struct S0
{
    public bool F1;
    public ushort F3;
    public byte F6;
}

public struct S1
{
    public int F0;
}

public class Program
{
    public static IRuntime s_rt;
    public static C1[][] s_1;
    public static S0[] s_3;
    public static S1 s_7;
    public static void Main()
    {
        ushort[] vr2 = default(ushort[]);
        S0 vr3 = default(S0);
        bool vr4 = vr3.F1;
        byte vr5 = vr3.F6--;
        if (vr4)
        {
            try
            {
                s_rt.WriteLine(1);
            }
            finally
            {
                C1 vr6 = s_1[0][0];
                var vr7 = new long[]{1};
                var vr8 = vr6.F3;
                long[] vr10 = vr7;
                C0 vr11 = vr8;
                int vr13 = s_7.F0;
                s_rt.WriteLine(vr6.F0.F0);
                s_rt.WriteLine(vr6.F0.F1);
                s_rt.WriteLine(vr6.F6.F0);
            }

            if (!vr4)
            {
                s_1[0][0].F3.F0 = vr3.F3;
            }
        }

        for (int vr9 = 0; vr9 < 2; vr9++)
        {
            if (vr4)
            {
                s_3[0].F3 = vr2[0];
            }

            s_rt.WriteLine(vr9);
        }
    }
}

public interface IRuntime
{
    void WriteLine<T>(T value);
}

public class Runtime : IRuntime
{
    public void WriteLine<T>(T value) => System.Console.WriteLine(value);
}
```

cc @EgorBo

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/71599

### CODE REVIEW

**Comment by EgorBo** [comment]
So the problem that `optLoopHoist` phase creates a new block [here](https://github.com/dotnet/runtime/blob/215d7e4ff4014755d3ebbb813c668ff6af8356a3/src/coreclr/jit/optimizer.cpp#L7933), it invalidates the assumption that `fgCurBBEpochSize == (fgBBNumMax + 1)`

`optJumpThreading` seems to not like it when it does `BlockSetOps::AddElemD` under the hood.

It seems to me that we can't just re-calculate doms and renumber blocks in-between loopHoist and optRedundantBranch phases (due to SSA state?)

so this fix works but it has some [diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1890015&view=ms.vss-build-web.run-extensions-tab). What do you think @AndyAyersMS @BruceForstall

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 1]
```diff
@@ -398,6 +398,12 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
                             break;
                         }
 
+                        if (fgCurBBEpochSize != (fgBBNumMax + 1))
+                        {
+                            // We added new blocks since the last renumerate e.g. in optLoopHoist
+                            break;
+                        }
+
```
Comment (AndyAyersMS) [inline src/coreclr/jit/redundantbranchopts.cpp:406]: I would put this check down at the start of `optJumpThread`.

**Review by AndyAyersMS — COMMENTED** [review]
Seems like an appropriate fix for now.

We really need to start eagerly creating preheaders, so hoisting can stop modifying the flow graph.

**Review by AndyAyersMS — APPROVED** [review]

**Comment by EgorBo** [comment]
Failure is https://github.com/dotnet/runtime/issues/72429

**Pull request merged**